thunder/transforms/quantization.py

```diff
@@ -0,0 +1,197 @@
+                absmax, shape=shape, blocksize=blocksize, code=quant_map, quant_type="nf4", dtype=dtype
+            )
+
+            return bitsandbytes.matmul_4bit(x, qweight.t(), bias=bias, quant_state=qs)
+
+        bnb_matmul_nf4 = bitsandbytes_executor.register_operator(
+            "bnb_matmul_nf4", meta=bnb_matmul_nf4_meta, fn=bnb_matmul_nf4_impl
+        )
+    return bitsandbytes_executor
+
+
+class BitsAndBytesLinearQuant4bit(EarlyTransform):
+    def __init__(self):
+        self.quant_states = {}
+        self.quantized_submodule_names = set()
+        get_bitsandbytes_executor()
+
+    def transform_module(self, model: thunder.ThunderModule):
+        self.thunder_module = model
+
+        def convert_linear_submodule(tm, name):
+            self.quantized_submodule_names.add(name)
+            weight_name = f"{name}.weight"
+            w = tm.get_parameter(weight_name)
+            # device!, double quant support
+            qw, qs = bitsandbytes.functional.quantize_4bit(w.to("cuda"), quant_type="nf4")
+            tm._overrides_parameters[weight_name] = qw
+            tm._overrides_parameters[f"{weight_name}.absmax"] = qs.absmax
+            tm._overrides_parameters[f"{weight_name}.code"] = qs.code
+            self.quant_states[weight_name] = {"dtype": qs.dtype, "shape": qs.shape, "blocksize": qs.blocksize}
+
+        for n, submodule in model._model.named_modules():
+            if isinstance(submodule, torch.nn.Linear):
+                convert_linear_submodule(model, n)
+
+    def transform_state_dict_for_submodule(self, model: thunder.ThunderModule, submodule_name: str, state_dict: dict):
+        # note that state dict entries do not include the submodule name as prefix
+        if submodule_name not in self.quantized_submodule_names:
+            return state_dict
+        weight_name_full = f"{submodule_name}.weight"
+        qs_dict = self.quant_states[weight_name_full]
+        w = state_dict["weight"]
+        assert w.dtype == qs_dict["dtype"]
+        assert w.shape == qs_dict["shape"]
+
+        qw, qs = bitsandbytes.functional.quantize_4bit(w.to("cuda"), block_size=qs_dict["blocksize"], quant_type="nf4")
+
+        # double quant support...
+        state_dict = state_dict.copy()
+        state_dict[weight_name] = qw
+        state_dict[f"{weight_name}.absmax"] = qs.absmax
+        state_dict[f"{weight_name}.code"] = qs.code
+
+        return state_dict
+
+    def transform_traces(self, prologue_trace, computation_trace, epilogue_trace, **kwargs):
+        tm = self.thunder_module
+        from thunder.core.trace import tracectx
+
+        checks = get_checks(prologue_trace)
+
+        compute_producers, compute_consumers = utils.producers_and_consumers(computation_trace)
+
+        if epilogue_trace is None:
+            prologue_to_epilogue_outputs = prologue_trace.output
+            output_subindex = None
+        else:
+            prologue_to_epilogue_outputs = prologue_trace.output[0]
```

Comment:
Just for my edification, can you explain these 4 lines?

Reply:
Ha, glad you asked after I filed #628 :) This is for traces with / without epilogues.

Reply:
I added a comment.